realtime/channel.py

```diff
@@ -3,69 +3,170 @@
+        self.timeout_task = None
+
+    async def resend(self):
+        self._cancel_ref_event()
+        self.ref = ""
+        self.ref_event = None
+        self.received_resp = None
+        self.sent = False
+        await self.send()
+
+    async def send(self):
+        if self._has_received("timeout"):
+            return
 
-        message = {
-            "topic": self.channel.topic,
-            "event": self.event,
-            "payload": self.payload,
-            "ref": self.ref,
-        }
+        self.start_timeout()
+        self.sent = True
 
         try:
-            await self.socket.ws_connection.send(json.dumps(message))
+            await self.channel.socket.send(
+                {
+                    "topic": self.channel.topic,
+                    "event": self.event,
+                    "payload": self.payload,
+                    "ref": self.ref,
+                    "join_ref": self.channel.join_push.ref,
+                }
+            )
         except Exception as e:
             logging.error(f"send push failed: {e}")
 
     def update_payload(self, payload: Dict[str, Any]):
         self.payload = {**self.payload, **payload}
 
+    def receive(self, status: str, callback: Callback) -> Push:
+        if self._has_received(status):
+            callback(self.received_resp.get("response", {}))
+
+        self.rec_hooks.append(Hook(status, callback))
+        return self
+
+    def start_timeout(self):
+        if self.timeout_task:
+            return
+
+        self.ref = self.channel.socket._make_ref()
+        self.ref_event = self.channel._reply_event_name(self.ref)
+
+        def on_reply(**kwargs):
+            self._cancel_ref_event()
+            self._cancel_timeout()
+            self.received_resp = kwargs.get("payload")
+            self._match_receive(**self.received_resp)
+
+        self.channel._on(self.ref_event, on_reply)
+
+        async def timeout(self):
+            await asyncio.sleep(self.timeout)
+            self.trigger("timeout", {})
+
+        self.timeout_task = asyncio.create_task(timeout(self))
+
+    def trigger(self, status: str, response: Any):
+        if self.ref_event:
+            payload = {
+                "status": status,
+                "response": response,
+            }
+            self.channel._trigger(self.ref_event, payload)
+
+    def destroy(self):
+        self._cancel_ref_event()
+        self._cancel_timeout()
+
+    def _cancel_ref_event(self):
+        if not self.ref_event:
+            return
+
+        self.channel._off(self.ref_event, {})
+
+    def _cancel_timeout(self):
+        if not self.timeout_task:
+            return
+
+        self.timeout_task.cancel()
+        self.timeout_task = None
+
+    def _match_receive(self, status: str, response: Any):
+        for hook in self.rec_hooks:
+            if hook.status == status:
+                # FIXME: callback can be a coroutine, how to handle that?
+                hook.callback(response)
```

Comment:
python experts: how to handle this?

The way it is, there are warnings saying the coroutine was never awaited.

Reply:
Just `await hook.callback(response)` ?

For more precise checking:

```python
import asyncio

def check_coroutine(callback):
    if asyncio.iscoroutine(callback):
        print("COROUTINE")
    else:
        print("NOT COROUTINE")

async def coroutine_function():
    pass

def normal_function():
    pass

check_coroutine(coroutine_function)
check_coroutine(normal_function)
```

Maybe code sample is better to understand.

Reply:
if I just `await hook.callback(response)` but it isn't a coroutine at runtime, wouldn't it fail/warn too?

Reply:
Edited it, see code sample.